### PR TITLE
GraphQL, JsonAPI and OpenApi specification endpoints

### PR DESCRIPTION
**Pull request by Jetski5822**

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Setup/GraphQL/GraphQLSchemaHashService.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+using System.Threading.Tasks;
+using OrchardCore.Apis.GraphQL;
+
+namespace OrchardCore.Contents.GraphQL
+{
+    public class GraphQLSchemaHashService : IGraphQLSchemaHashService
```
Comment (rserj) [inline src/OrchardCore.Modules/OrchardCore.Setup/GraphQL/GraphQLSchemaHashService.cs]: Please see my comments in OrchardCore.Apis.GraphQL/Services/SchemaService.cs regarding to IGraphQLSchemaHashService

Comment (Jetski5822) [reply src/OrchardCore.Modules/OrchardCore.Setup/GraphQL/GraphQLSchemaHashService.cs]: argh!! Cant find your comments?!?!?!? any ideas how to get to them?

**Comment by Jaben** [comment]
Any update on this PR? Would like to test Orchard Core as a headless CMS.

**Comment by Jetski5822** [comment]
@Jaben not dead - Just busy :) - There is one test I need to fix, and it would appear a merge too -and then it should be good.

File: src/OrchardCore/OrchardCore.Application.Cms.Targets/ServiceExtensions.cs [hunk 2]
```diff
@@ -1,38 +1,38 @@
...
-                    "OrchardCore.Antiforgery",
-					"OrchardCore.Mvc", 
-                    "OrchardCore.Settings", 
-                    "OrchardCore.Setup",
-                    "OrchardCore.Recipes", 
-                    "OrchardCore.Commons", 
-                    "OrchardCore.Apis.GraphQL",
+                .AddSetupFeatures(
+					"OrchardCore.Setup",
+					"OrchardCore.Apis.GraphQL",
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Application.Cms.Targets/ServiceExtensions.cs]: Nope

File: src/OrchardCore/OrchardCore.Application.Cms.Targets/ServiceExtensions.cs [hunk 3]
```diff
@@ -16,7 +16,11 @@ public static IServiceCollection AddOrchardCms(this IServiceCollection services)
                 .AddAuthentication()
                 .AddDataProtection()
 
-                .AddSetupFeatures("OrchardCore.Setup")
+                .AddSetupFeatures(
+					"OrchardCore.Setup",
+					"OrchardCore.Apis.GraphQL",
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Application.Cms.Targets/ServiceExtensions.cs]: Nope. Recipe only, only if we want it by default, which I don't think so for the recipes we have today.

Comment (sebastienros) [reply src/OrchardCore/OrchardCore.Application.Cms.Targets/ServiceExtensions.cs]: In recipes, yes (again if we had such a recipe), not hard coded in the host application.

Comment (Jetski5822) [reply src/OrchardCore/OrchardCore.Application.Cms.Targets/ServiceExtensions.cs]: Shifting to the Setup module

File: test/OrchardCore.Tests/Apis/SiteStartup.cs [hunk 4]
```diff
@@ -22,11 +21,17 @@ public SiteStartup(TestSiteConfiguration configuration)
 
         public void ConfigureServices(IServiceCollection services)
         {
-            services.AddOrchardCms();
+            services.AddOrchardCms().AddOrchardCore()
```
Comment (sebastienros) [inline test/OrchardCore.Tests/Apis/SiteStartup.cs]: If we need to call `AddOrchardCore` after `AddOrchardCms` to be able to configure the features, then we need to add an overload to accept an `Action<OrchardCoreBuilder>` in `AddOrchardCms`.

Comment (jtkech) [reply test/OrchardCore.Tests/Apis/SiteStartup.cs]: `AddOrchardCore()` has this ability.

Why not just letting `AddOrchardCms()` returning a builder as `AddOrchardCore()` do?

Comment (sebastienros) [reply test/OrchardCore.Tests/Apis/SiteStartup.cs]: It _could_ return an `OrchardCmsBuilder` but it should still accept an `OrchardCoreBuilder` as a custom action.

Comment (jtkech) [reply test/OrchardCore.Tests/Apis/SiteStartup.cs]: Okay cool, as `AddOrchardCore()`, i will update it.

Comment (jtkech) [reply test/OrchardCore.Tests/Apis/SiteStartup.cs]: For infos.

Just saw that the overload of `AddOrchardCore()` which accepts an action as been removed. But maybe useless here.

**Comment by xperiandri** [comment]
@Jetski5822, can I help you with this PR somehow?

**Comment by xperiandri** [comment]
@Jetski5822, thanks for the update! I will try it.
Could you write or record or point to a recorded demo of what's available?

**Comment by Jetski5822** [comment]
hey @xperiandri - I just need to finish off the client... I have some stuff locally I need to polish off.

The GraphQL and OpenAPI stuff I am happy with.. its just the JsonAPI part I am finishing off.

**Comment by jtkech** [comment]
Just retried, tests fail if run all together, but they pass if run individually.

- Seems that the `ContentRootPath` is not always correctly set in the testing environmment, so as a temporary fix i added a null check. Then all tests passed but needs a better fix.

**Update**: hmm, still failling but all is ok on my machine, , i will check it tomorrow.

- I used the new `AddOrchardCms()` version which accepts a configure action.

- In `OC.Tests` you reference `OC.Cms.Web`, looks strange for me but if we do this many project references become useless, so i cleaned up them.

- An `App_Data` folder is also created under `test\OrchardCore.Tests\bin\Debug\netcoreapp2.1\` and under its `logs` subfolder there is a `internal-nlog.txt` with some nlog errors.

- There are a bunch of `Sites_random_number` folders created under `src\OrchardCore.Cms.Web\App_Data`. E.g `Sites_0a55a1b976204883b6df6155297de1e7`. We would need to clean them automatically.

- There are some lucene writer locking issues but it is another issue that might be fixed by #1669.

**Comment by jtkech** [comment]
Some progress, just failing on appveyor but again all tests pass on my machine and now even if i run them all together (see my comments above).

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            _hashService = hashService;
+            _serviceProvider = serviceProvider;
+        }
+
+        public async Task<ISchema> GetSchema()
+        {
+            var hash = await _hashService.GetHash();
+
+            if (_lastHash != hash)
+            {
```
[the remark this answers is not in the document]
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: Please use a IDistributedCache if you intend to do so. It's already defaulting to it's own IMemoryCache implementation.

**Comment by Skrypt** [comment]
I tested this branch and first thing I noticed is that the GraphiQL editor should be better integrated. It should take all the space in that right content panel without adding a Title to that page.

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        public static void AddGraphQueryType<T>(this IServiceCollection services) where T : QueryFieldType
+        {
+            services.AddScoped<T>();
+            services.AddScoped<QueryFieldType, T>();
+        }
+
+        public static void AddGraphQLInputType<TContenPart, TInputType>(this IServiceCollection services) 
+            where TContenPart : class 
+            where TInputType : InputObjectGraphType<TContenPart>
+        {
```
Comment (rserj) [inline src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs]: Probably we should rename TContenPart to something else, cause `OrchardCore.Apis.GraphQL.Abstractions` project, should not be CMS specific

**Comment by jtkech** [comment]
@Jetski5822 thanks for using my PR.

Just as a reminder

- To make `RecentBlogPostsQueryTests` working i needed to fix some lucene writer locking issues by using `ShellContext.cs` and `ShellHost.cs` from #1669. But they are part of another PR, so maybe better to revert this and still skip this test temporarily, i don't know, just to be aware of this.

**Comment by sebastienros** [comment]
I will add my review very soon, can you answer the other comments in the meantime?

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs (hunk 6 above)
Comment (Jetski5822) [reply src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs]: Agreed! - Resolved

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jetski5822) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs]: The hash system was something @sebastienros used elsewhere.... Seb - is my implementation correct?

**Comment by Jetski5822** [comment]
@sebastienros how are those review comments looking?

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            {
+                body = await streamReader.ReadToEndAsync().ConfigureAwait(true);
+            }
+
+            var request = JsonConvert.DeserializeObject<GraphQLRequest>(body);
+
+            var queryToExecute = request.Query;
+            if (!String.IsNullOrWhiteSpace(request.NamedQuery))
+            {
+                var namedQueries = context.RequestServices.GetServices<INamedQueryProvider>();
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs]: This is very un-optimal as it creating new dictionaries for every single call. I suggest to change `INamedQueryProvider` to have a `Recolve` method that return the query directly. Do not use dictionaries here. The first implementation for return a result wins.

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        private async Task ExecuteAsync(HttpContext context, ISchemaService schemaService)
+        {
+            var schema = await schemaService.GetSchema();
+
+            string body;
+            using (var streamReader = new StreamReader(context.Request.Body))
+            {
+                body = await streamReader.ReadToEndAsync().ConfigureAwait(true);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs]: Remove `ConfigureAwait`, and do it everywhere if it's not the only one.

**Comment by sebastienros** [comment]
The performance is really problematic with my simple tests. I get 60 RPS locally to get the blog posts (a single one) versus 1K RPS to render the home page that also loads all blog posts.

Sample query:

```
{
	query: "{
		blogPost(published: true) {
    		titlePart {
    			title
		    },
			owner   
		}
	}"
}
```

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Client/ContentResource.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            var contentTypeBuilder = new ContentTypeQueryResourceBuilder(contentType);
+            builder(contentTypeBuilder);
+
+            var requestJson = new JObject(
+                new JProperty("query", @"query { " + contentTypeBuilder.Build() + " }")
+                );
+
+            var response = await _client
+                .PostJsonAsync("api/graphql", requestJson.ToString());
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Apis.GraphQL.Client/ContentResource.cs]: `"api/graphql"` is a setting, it can't be ahrd coded. Ditto for other usages.

**Comment by jtkech** [comment]
@sebastienros 

So, since we use memory cache, services cached through `ContentSchema` need to be singletons or transients but not scoped (or new objects created e.g through factories). I made them singletons but some are doing things through their constructors which need to be re-run, so i made these ones transients. Another way would be to explicitly create new objects or use factories by still using the DI.

So, i have a version where all unit tests pass but i will not push it because it is now conflicting with the current version. But i think you already got the idea, so i will let you do the stuff.

And if you have some problems with unit tests i will restart from your latest version.

**Comment by jtkech** [comment]
@sebastienros for infos.

So, i could get all unit tests working by only using transient services for graphql related types so that they can be cached and some act as "factories" through their ctors when resolved.

For infos, it was not working for only one generic type. I could fix it in `ContentItemFieldTypeProvider` by doing this little update.

    var queryGraphType =
        typeof(ObjectGraphType<>).MakeGenericType(activator.Type);

    ...

    var queryGraphTypeResolved =  (IObjectGraphType)_httpContextAccessor.HttpContext
        .RequestServices.GetService(queryGraphType);

    if (queryGraphTypeResolved != null)
    {
        typeType.Field(
            // *** I NEEDED TO REPLACE THIS LINE ***
            // queryGraphType,
            // *** WITH THIS LINE ***
            queryGraphTypeResolved.GetType(),
            partName,
            resolve: context =>
            {
                ...
            });
        }

So, i mainly kept what @Jetski5822 did and just because of the schema caching i mainly replaced scoped services with transients which use `IHttpContextAccessor`. And the above update for unit tests.

Note: i made `IDependencyResolver` a singleton (which also uses `IHttpContextAccessor`) but it could also be transient, and `GraphQLSchemaHashService` and `IApiUpdateModel` are still scoped services.

I will make a testing branch just to show and target dev just to see if unit tests pass.

File: src/OrchardCore/OrchardCore.Apis.GraphQL.Client/ContentResource.cs (hunk 9 above)
Comment (carlwoodhouse) [reply src/OrchardCore/OrchardCore.Apis.GraphQL.Client/ContentResource.cs]: hah @Jetski5822 - this is what i was on about earlier ;)

**Comment by sebastienros** [comment]
They released a new version of GraphQL.NET that fixes the lifetime issues we described. We'll need to refactor this PR to handle that.

**Comment by obfu5c8** [comment]
Tried out a wildcard query - looks like it doesn't filter at all - just returns the entire dataset :(

```
{
  patientInformationLeaflet(TitlePart:{ title:"%throat%"}){
    titlePart {
      title
    }
    contentItemId
  }
}
```

**Comment by sebastienros** [comment]
@obfu5c8 looks correct. Where did you see that you could pass a property like this as a parameter? Or did you create a `title` parameter in the query itself, in which case the syntax is not correct, and the value would be escaped which would result in an empty result.

**Comment by Jetski5822** [comment]
The query is correct, looking at the input type for titlePart and not sure this is actually allowed.... will dig in to it tomorrow

**Comment by Jetski5822** [comment]
need to look in to authorization https://github.com/graphql-dotnet/authorization/blob/master/src/Harness/GraphQL.cs

**Comment by jtkech** [comment]
@Jetski5822 

Also in `CreateTenantMutation`, when you get errors after running a setup, maybe good to add the following as now done in a regular setup or through a `JsonApi`. Otherwise if the setup fails you will not be able to do another request on this tenant, in this case we now return a 503.

    shellSettings.State = TenantState.Uninitialized;

Not thread safe but it is not allowed to run 2 setups concurrently on the same tenant. We now return a 503 if the tenant has started its initialization, but it is still possible to trigger 2 setups concurrently.

File: src/OrchardCore.Modules/OrchardCore.Setup/GraphQL/CreateTenantMutation.cs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            _httpContextAccessor = httpContextAccessor;
+            Name = "CreateTenant";
+
+            Arguments = new AutoRegisteringQueryArguments<SetupViewModel>();
+
+            Type = typeof(CreateTenantOutcomeType);
+
+            Resolver = new AsyncFieldResolver<object, object>(async (context) =>
+            {
+                var serviceProvider = httpContextAccessor.HttpContext.RequestServices;
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Setup/GraphQL/CreateTenantMutation.cs]: maybe better to use the field `_httpContextAccessor` initialized above, or just not declare it.

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Mutations/Types/CreateContentItemInputType.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+            foreach (var part in contentParts)
+            {
+                var partName = part.GetType().Name;
+                var activator = typeActivator.GetTypeActivator(partName);
+
+                var inputGraphType =
+                    typeof(InputObjectGraphType<>).MakeGenericType(activator.Type);
+
+                var inputGraphTypeResolved = (IInputObjectGraphType)httpContextAccessor.HttpContext
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Mutations/Types/CreateContentItemInputType.cs]: maybe better to use the field `_httpContextAccessor` if you initialize it before (not at the end), or just not declare it.

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Mutations/DeleteContentItemMutation.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+                new QueryArgument<NonNullGraphType<StringGraphType>> { Name = "contentItemId" }
+            );
+
+            Type = typeof(DeletionStatusObjectGraphType);
+
+            Resolver = new AsyncFieldResolver<object, DeletionStatus>(async (context) =>
+            {
+                var contentItemId = context.GetArgument<string>("contentItemId");
+
+                var contentManager = httpContextAccessor.HttpContext.RequestServices.GetRequiredService<IContentManager>();
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Mutations/DeleteContentItemMutation.cs]: maybe better to use the field `_httpContextAccessor` if you initialize it before (not at the end), or just not declare it.

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Mutations/DeleteContentItemMutation.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
+using System.Threading.Tasks;
+using GraphQL.Resolvers;
+using GraphQL.Types;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Apis.GraphQL;
+using OrchardCore.Apis.GraphQL.Types;
+using OrchardCore.ContentManagement;
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Mutations/DeleteContentItemMutation.cs]: Remove unused `using`.

File: src/OrchardCore.Modules/OrchardCore.Setup/Controllers/SetupController.cs [hunk 14]
```diff
@@ -74,6 +74,62 @@ public async Task<ActionResult> Index()
...
+
+            // Check if a component in the Setup failed
+            if (setupContext.Errors.Any())
+            {
+                foreach (var error in setupContext.Errors)
+                {
+                    ModelState.AddModelError(error.Key, error.Value);
+                }
+
+                _shellSettings.State = TenantState.Uninitialized;
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Setup/Controllers/SetupController.cs]: you will not need this if #2284 is merged.

File: src/OrchardCore.Modules/OrchardCore.Setup/Controllers/SetupController.cs [hunk 15]
```diff
@@ -74,6 +74,62 @@ public async Task<ActionResult> Index()
...
+
+            // Check if a component in the Setup failed
+            if (setupContext.Errors.Any())
+            {
+                foreach (var error in setupContext.Errors)
+                {
+                    ModelState.AddModelError(error.Key, error.Value);
+                }
+		
+                _shellSettings.State = TenantState.Uninitialized;
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Setup/Controllers/SetupController.cs]: you will not need this if #2284 is merged.

File: src/OrchardCore.Modules/OrchardCore.Lucene/LuceneIndexManager.cs [hunk 16]
```diff
@@ -260,6 +260,12 @@ private void Write(string indexName, Action<IndexWriter> action, bool close = fa
                     {
                         var directory = CreateDirectory(indexName);
                         var analyzer = _luceneAnalyzerManager.CreateAnalyzer(LuceneSettings.StandardAnalyzer);
+
+                        if (analyzer == null) {
+                            _luceneAnalyzerManager.RegisterAnalyzer(new LuceneAnalyzer(LuceneSettings.StandardAnalyzer, new StandardAnalyzer(LuceneSettings.DefaultVersion)));
+                            analyzer = _luceneAnalyzerManager.CreateAnalyzer(LuceneSettings.StandardAnalyzer);
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Lucene/LuceneIndexManager.cs]: Maybe you will not need this anymore if #2285 is merged.

**Comment by agriffard** [comment]
https://github.com/graphql/graphiql/releases updated to 0.12

**Comment by Skrypt** [comment]
I suggest that we expose the ReDoc documentation but this could be done in a separate PR.

**Comment by jrestall** [comment]
What items need to be completed on this PR so that we can at least get it into dev?

**Comment by Skrypt** [comment]
I started merging dev branch on top of it on my own branch (not completed yet). And we need to separate the JSON Api stuff and GraphQL on 2 separate PR's. We need to start an issue with what's required for only making a JSON Api that will work with modules too.

**Comment by jtkech** [comment]
@Jetski5822 

> Dont pull just yet as my code is pointing at a local GraphQL project

Yes, that's why i needed to do this here #2497 to be able to build locally. See also some little csproj cleanup and corrections, e.g when using the `Sdk.Razor` we need to reference `AspNetCore.Mvc` directly.

**Comment by Jetski5822** [comment]
The JSONApi stuff should just work... will try tomorrow

**Comment by jtkech** [comment]
To make unit tests working on my machine i needed to make the text field graph type nullable. This to not fail on a query when a property is not yet defined on an item. Maybe we need to do it for other fields.

I committed this little change directly on your branch.

**Comment by jrestall** [comment]
I've added some feedback at issue https://github.com/OrchardCMS/OrchardCore/issues/2360#issuecomment-429665633 that discusses the content fields query structure. 

@jtkech I made all fields nullable on my branch as I hit the same issue. You can see [here](https://github.com/jrestall/OrchardCore/commit/1c44dc6ee09348f0257ef5fdb1a6b84dee76adb6#diff-ed8b0d40a8ba97ee6d3e76d3605d32e9).

**Comment by jtkech** [comment]
@jrestall okay cool, i let @Jetski5822 doing it for the others.

At least he will be happy to see that all is green ;)

**Comment by Jetski5822** [comment]
Totally cool man - go for it. I have a few changes to make too :)

**Comment by sebastienros** [comment]
Closing this PR, not the branch as we need its content.

**Pull request closed without merging**